cpp/src/arrow/sparse_tensor.cc

```diff
@@ -301,6 +303,130 @@ class SparseTensorConverter<TYPE, SparseCSRIndex>
   inline Status CheckMaximumValue(const uint64_t) const { return Status::OK(); }
 };
 
+// ----------------------------------------------------------------------
+// SparseTensorConverter for SparseCSFIndex
+
+template <typename TYPE>
+class SparseTensorConverter<TYPE, SparseCSFIndex>
+    : private SparseTensorConverterBase<TYPE> {
+ public:
+  using BaseClass = SparseTensorConverterBase<TYPE>;
+  using typename BaseClass::NumericTensorType;
+  using typename BaseClass::value_type;
+
+  SparseTensorConverter(const NumericTensorType& tensor,
+                        const std::shared_ptr<DataType>& index_value_type,
+                        MemoryPool* pool)
+      : BaseClass(tensor, index_value_type, pool) {}
+
+  template <typename IndexValueType>
+  Status Convert() {
+    using c_index_value_type = typename IndexValueType::c_type;
+    RETURN_NOT_OK(CheckMaximumValue(std::numeric_limits<c_index_value_type>::max()));
+
+    std::shared_ptr<SparseCOOTensor> sparse_coo_tensor;
+    ARROW_ASSIGN_OR_RAISE(sparse_coo_tensor, SparseCOOTensor::Make(tensor_));
+    std::shared_ptr<Tensor> coords =
+        arrow::internal::checked_pointer_cast<SparseCOOIndex>(
+            sparse_coo_tensor->sparse_index())
+            ->indices();
+
+    // Convert SparseCOOTensor to long CSF buffers
+    const int64_t ndim = tensor_.ndim();
+    const int64_t nonzero_count = sparse_coo_tensor->non_zero_length();
+
+    std::vector<int64_t> counts(ndim);
+    std::fill_n(counts.begin(), ndim, static_cast<int64_t>(0));
+    std::vector<int64_t> axis_order = internal::ArgSort(tensor_.shape());
+
+    std::vector<TypedBufferBuilder<c_index_value_type>> indptr_buffer_builders(ndim - 1);
+    std::vector<TypedBufferBuilder<c_index_value_type>> indices_buffer_builders(ndim);
```

Comment:
Should we [reserve](https://arrow.apache.org/docs/cpp/api/memory.html#_CPPv4N5arrow13BufferBuilder7ReserveEK7int64_t) `TypedBufferBuilder` here? We can't really know in advance, maybe we should approximate?